dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,256 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Control.Exception (handle, SomeException(..))
+import Control.Monad (foldM)
+import Data.List (isSuffixOf)
+import Data.Maybe (fromMaybe)
+import Data.Semigroup (Semigroup(..))
+import Dhall.Map (foldMapWithKey)
+import Data.Text (Text)
+import Data.Text.Encoding (encodeUtf8)
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(srcStart))
+import Dhall.Parser (exprFromText)
+import System.Directory ( doesDirectoryExist
+#if MIN_VERSION_directory(1,3,0)
+                        , pathIsSymbolicLink
+#elif MIN_VERSION_directory(1,2,6)
+                        , isSymbolicLink
+#endif
+                        , getDirectoryContents
+                        )
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+import qualified Data.ByteString as BS (length)
+import qualified Data.Map      as M
+import qualified Data.Text     as T
+import qualified Data.Text.IO  as TIO
+
+{-| 
+    Documentation for ETags format is not very informative and not very correct.
+    You can find version of documentation here:
+    https://en.wikipedia.org/wiki/Ctags#Etags_2
+    and you can also check source code here:
+    http://cvs.savannah.gnu.org/viewvc/vtags/vtags/vtags.el?view=markup
+-}
+
+data PosInFile = FP
+    { fpLine :: Int
+      -- ^ line number, starting from 1, whe to find tag
+    , fpOffset :: Int
+      -- ^ byte offset form start of file. Not sure if any editor use it
+    } deriving (Eq, Ord)
+
+newtype Tags = Tags (M.Map FilePath [(PosInFile, Tag)])
+
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith (<>) ts1 ts2)
+
+instance Monoid Tags where
+    mempty = Tags M.empty
+    mappend = (<>)
+
+{-| For example, for line: `let foo = "foo"` tag is:
+    `Tag "let " "foo"`
```

Comment:
Can you check what the haddocks for this look like, e.g. with `stack haddock --open dhall`?

I suspect something like this would be better

```
> let foo = "bar"

> Tag "let " "foo"
```

Reply:
Never mind. If we only export `generate`, we don't get any haddocks here.

Reply:
I still can fix it. So if someone will need something except generate in future.